workflow/engine.py

```diff
@@ -391,315 +401,545 @@ def run_callbacks(self, callbacks, objects, obj, indent=0):
+        * "first": first object
+
+        `task` must be either:
+
+        * "prev": previous task
+        * "current": current task
+        * "next": next task
+        * "first": first task
+
+        To continue with next object from the first task:
+
+        .. code-block:: python
+
+                wfe.restart("next", "first")
+
+        :param obj: the object which should be restarted
+        :type obj: str
+
+        :param task: the task which should be restarted
+        :type task: str
+        """
+        if not self.callbacks._dict:
+            raise Exception("The callbacks are empty, did you set workflows?")
+
+        self.log.debug("Restarting workflow from {0} object and {1} task"
+                       .format(str(obj), str(task)))
+
+        # set the token from which to start processing
+        if obj == 'prev':
+            # start with the previous object
+            self.state.elem_ptr -= 2
+        elif obj == 'current':
+            # continue with the current object
+            self.state.elem_ptr -= 1
+        elif obj == 'next':
+            pass
+        elif obj == 'first':
+            self.state.elem_ptr = 0
         else:
-            return self._callbacks
+            raise Exception('Unknown start point for object: %s' % obj)
+
+        # set the task that will be executed first
+        if task == 'prev':
+            # the previous
+            self.state.task_pos[-1] -= 1
+        elif task == 'current':
+            # restart the task again
+            pass
+        elif task == 'next':
+            # continue with the next task
+            self.state.task_pos[-1] += 1
+        elif task == 'first':
+            self.state.task_pos = [0]
+        else:
+            raise Exception('Unknown start point for task: %s' % obj)
+
+        if objects:
+            self._process(objects)
+        else:
+            self._process(self._objects)
+
+    # XXX Now a property
+    # XXX Now returns only active objects
+    @property
+    def objects(self):
+        """Return iterator for walking through the objects."""
+        return (obj for obj in self._objects)
+
+    @property
+    def current_object(self):
+        """Return the currently active DbWorkflowObject."""
+        if self.state.elem_ptr < 0:
+            return None
+        return self._objects[self.state.elem_ptr]
+
+    @property
+    def has_completed(self):
+        """Return whether the engine has completed its execution."""
+        if self.state.elem_ptr == -1:
+            return False
+        return len(self._objects) - 1 == self.state.elem_ptr and \
+            self._current_object_processed
+
+    # Deprecated
+    @property
+    def store(self):
+        return self.extra_data
+
+    # Deprecated
+    def setWorkflow(self, list_or_tuple):
+        return self.set_callbacks(list_or_tuple)
+
+    # Deprecated
+    def setPosition(self, elem_ptr, task_pos):
+        # """Set the internal pointers (of current state/obj).
```

Comment:
This is intentional as it prevents sphinx from adding the method to the API listing.